### PR TITLE
feat: fix 404 on the first app loads when unauthenticated

### LINKED ISSUE #9958: `404` first time with shared path-based `coder_app` even with `CODER_DANGEROUS_ALLOW_PATH_APP_SHARING` set to true
Version: [`v2.2.0+d756e93`](https://github.com/coder/coder/commit/d756e93469d7fdaef831581aaad0cdc480b85c05)

1. My deployment (Kubernetes) is set to allow path app sharing

```yaml
# allow authenticated users to access path-based workspace apps
    - name: CODER_DANGEROUS_ALLOW_PATH_APP_SHARING
      value: "true"
```

2. I start a workspace with `coder_app` `share=authenticated` and the pretty globe shows up next to path-based `code-server`

image

3. In another browser, I paste the `code-server` URL and it takes me to the Coder deployment's authentication page - this is expected behavior
4. After successfully authenticated to Coder as a user with `Member` role, I get a `404`

image

5. If I then refresh the browser UI, I'm in `code-server`

image

I also tested this with an `Owner` role and it happened.

### Resources

Related, closed fixed issue #9016 but I think was if `CODER_DANGEROUS_ALLOW_PATH_APP_SHARING` was not set to `true`

@Emyrk @matifali 

Additional testing across browsers, makes me think cookies are involved. i.e., I would forcibly log out of Coder, but paste the code-server URL, and it would connect, and share=authenticated

### PR DESCRIPTION
**Pull request by Emyrk**
Closes: https://github.com/coder/coder/issues/9958

# What the issue was

1. Hit app page **without any cookies**
2. Get redirected with a 303 to `/login` with `redirect` query param set to the app url.
3. Hit the login page
4. Enter credentials
5. [This code](https://github.com/coder/coder/blob/stevenmasley%2Fautostart_schedule/site/src/pages/LoginPage/LoginPage.tsx#L24) navigates you to the redirect url
6. You hit a 404 page

The technical thing to note is that step 5 is doing a react navigation. We need to send this URL in an HTTP request to the Golang backend to generate the `coder_signed_app_token` cookie. The react `Navigate` does not send an HTTP request to the `redirectTo` url.

This is also an issue with subdomain based apps. The redirect fails.

# What changes

This changes all app based redirects to a window.location.href change. This triggers a new HTTP request to the redirect url which the Golang BE can return a `set-cookie` for the app authentication.

### CODE REVIEW

File: site/src/pages/LoginPage/LoginPage.tsx [hunk 1]
```diff
@@ -21,6 +21,27 @@ export const LoginPage: FC = () => {
...
+        const redirectURL = new URL(redirectTo);
+        if (redirectURL.host !== window.location.host) {
+          window.location.href = redirectTo;
+        }
+      } catch {
+        // Do nothing
+      }
+      // Path based apps.
+      if (redirectTo.includes("/apps/")) {
+        window.location.href = redirectTo;
```
Comment (aslilac) [inline site/src/pages/LoginPage/LoginPage.tsx:42]:
```suggestion
          window.location.href = redirectTo;
          return;
        }
      } catch {
        // Do nothing
      }
      // Path based apps.
      if (redirectTo.includes("/apps/")) {
        window.location.href = redirectTo;
        return;
```

Comment (aslilac) [reply site/src/pages/LoginPage/LoginPage.tsx:42]: need `return` after these assignments because, very weirdly, this specific assignment is async 🤪

Comment (Emyrk) [reply site/src/pages/LoginPage/LoginPage.tsx]: @aslilac aside from that, can we merge this?

**Review by BrunoQuaresma — APPROVED** [review]
Thanks for adding the comments! I don't like this way too much but I can see why it is needed... maybe in the future we can figure out a better solution.

**Pull request merged**